src/ring_theory/perfection.lean

```diff
@@ -0,0 +1,366 @@
+
+/-- The perfection of a ring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def ring.perfection (R : Type u) [comm_ring R] (p : ℕ) [hp : fact p.prime] [char_p R p] :
+  subring (ℕ → R) :=
+{ neg_mem' := λ f hf n, (frobenius_neg R p _).trans $ congr_arg _ (hf n),
+  .. comm_semiring.perfection R p }
+
+namespace ring.perfection
+
+variables (R : Type u) [comm_ring R] (p : ℕ) [hp : fact p.prime] [char_p R p]
+include hp
+
+/-- The `n`-th coefficient of an element of the perfection. -/
+def coeff (n : ℕ) : ring.perfection R p →+* R :=
+{ to_fun := λ f, f.1 n,
+  map_one' := rfl,
+  map_mul' := λ f g, rfl,
+  map_zero' := rfl,
+  map_add' := λ f g, rfl }
+
+variables {R p}
+
+@[ext] lemma ext {f g : ring.perfection R p} (h : ∀ n, coeff R p n f = coeff R p n g) : f = g :=
+subtype.eq $ funext h
+
+variables (R p)
+
+/-- The `p`-th root of an element of the perfection. -/
+def pth_root : ring.perfection R p →+* ring.perfection R p :=
+{ to_fun := λ f, ⟨λ n, coeff R p (n + 1) f, λ n, f.2 _⟩,
+  map_one' := rfl,
+  map_mul' := λ f g, rfl,
+  map_zero' := rfl,
+  map_add' := λ f g, rfl }
+
+variables {R p}
+
+lemma coeff_pth_root (f : ring.perfection R p) (n : ℕ) :
+  coeff R p n (pth_root R p f) = coeff R p (n + 1) f :=
+rfl
+
+lemma coeff_pow_p (f : ring.perfection R p) (n : ℕ) :
+  coeff R p (n + 1) (f ^ p) = coeff R p n f :=
+by { rw ring_hom.map_pow, exact f.2 n }
+
+lemma coeff_frobenius (f : ring.perfection R p) (n : ℕ) :
+  coeff R p (n + 1) (frobenius _ p f) = coeff R p n f :=
+by convert coeff_pow_p f n
+
+lemma pth_root_frobenius : (pth_root R p).comp (frobenius _ p) = ring_hom.id _ :=
+ring_hom.ext $ λ x, ext $ λ n,
+by rw [ring_hom.comp_apply, ring_hom.id_apply, coeff_pth_root, coeff_frobenius]
+
+lemma frobenius_pth_root : (frobenius _ p).comp (pth_root R p) = ring_hom.id _ :=
+ring_hom.ext $ λ x, ext $ λ n,
+by rw [ring_hom.comp_apply, ring_hom.id_apply, ring_hom.map_frobenius, coeff_pth_root,
+    ← ring_hom.map_frobenius, coeff_frobenius]
+
+lemma coeff_add_ne_zero {f : ring.perfection R p} {n : ℕ} (hfn : coeff R p n f ≠ 0) (k : ℕ) :
+  coeff R p (n + k) f ≠ 0 :=
+nat.rec_on k hfn $ λ k ih h, ih $ by erw [← coeff_pow_p, ring_hom.map_pow, h, zero_pow hp.pos]
+
+lemma coeff_ne_zero_of_le {f : ring.perfection R p} {m n : ℕ} (hfm : coeff R p m f ≠ 0)
+  (hmn : m ≤ n) : coeff R p n f ≠ 0 :=
+let ⟨k, hk⟩ := nat.exists_eq_add_of_le hmn in hk.symm ▸ coeff_add_ne_zero hfm k
+
+end ring.perfection
+
+section perfectoid
+
+variables (K : Type u) [field K] (v : valuation K ℝ≥0)
+variables (O : Type v) [comm_ring O] [algebra O K] (hv : v.integers O)
+variables (p : ℕ)
+include hv
+
+/-- `O/(p)` for `O`, ring of integers of `K`. -/
+@[nolint unused_arguments has_inhabited_instance] def mod_p :=
+(ideal.span {p} : ideal O).quotient
+
+variables [hp : fact p.prime] [hvp : fact (v p ≠ 1)]
```

Comment:
@digama0 @cipher1024 @gebner Do I understand correctly that `fact p.prime` was a good idea that I maybe executed badly. Is `[fact (v p \ne 1)]` ok?? Or is this going to make the type class system sad?
If so, what is the best solution?

Reply:
I *think* that a hypothesis `[fact P]` is ok for almost any proposition `P` (as long as enough lemmas want the same hypothesis). There should just not be any global instances of the form `fact ?m` (which, according to https://leanprover-community.github.io/mathlib_docs/logic/basic.html#fact means that there are 9 too much global instances, IMO)

Reply:
Ok, then let's not do anything here (-;